src/extractor/way_restriction_map.cpp

```diff
@@ -0,0 +1,199 @@
+                way.in_restriction.to == way.out_restriction.via)
+                restriction_data.push_back(turn_restriction);
+        }
+    };
+    std::for_each(turn_restrictions.begin(), turn_restrictions.end(), extract_restrictions);
+
+    const auto as_duplicated_node =
+        [](auto const &restriction) -> std::tuple<NodeID, NodeID, NodeID> {
+        auto &way = restriction.AsWayRestriction();
+        // group restrictions by the via-way. On same via-ways group by from
+        return std::make_tuple(
+            way.in_restriction.via, way.out_restriction.via, way.in_restriction.from);
+    };
+
+    const auto by_duplicated_node = [&](auto const &lhs, auto const &rhs) {
+        return as_duplicated_node(lhs) < as_duplicated_node(rhs);
+    };
+
+    std::sort(restriction_data.begin(), restriction_data.end(), by_duplicated_node);
+
+    std::size_t index = 0, duplication_id = 0;
+    // map all way restrictions into access containers
+    const auto prepare_way_restriction = [this, &index, &duplication_id, as_duplicated_node](
+        const auto &restriction) {
+        const auto &way = restriction.AsWayRestriction();
+        restriction_starts.insert(
+            std::make_pair(std::make_pair(way.in_restriction.from, way.in_restriction.via), index));
+        ++index;
+    };
+    std::for_each(restriction_data.begin(), restriction_data.end(), prepare_way_restriction);
+
+    std::size_t offset = 1;
+    // the first group starts at 0
+    if (!restriction_data.empty())
+        duplicated_node_groups.push_back(0);
+
+    auto const add_offset_on_new_groups = [&](auto const &lhs, auto const &rhs) {
+        BOOST_ASSERT(rhs == restriction_data[offset]);
+        // add a new lower bound for rhs
+        if (as_duplicated_node(lhs) != as_duplicated_node(rhs))
+            duplicated_node_groups.push_back(offset);
+        ++offset;
+        return false; // continue until the end
+    };
+    std::adjacent_find(restriction_data.begin(), restriction_data.end(), add_offset_on_new_groups);
+    duplicated_node_groups.push_back(restriction_data.size());
+}
+
+std::size_t WayRestrictionMap::NumberOfDuplicatedNodes() const
+{
+    return duplicated_node_groups.size() - 1;
+}
+
+bool WayRestrictionMap::IsViaWay(const NodeID from, const NodeID to) const
+{
+    // safe-guards
+    if (restriction_data.empty())
+        return false;
+
+    const auto itr = std::lower_bound(
+        restriction_data.begin(), restriction_data.end(), std::make_tuple(from, to), FindViaWay());
+
+    // no fitting restriction
+    if (itr == restriction_data.end())
+        return false;
+
+    const auto &way = itr->AsWayRestriction();
+
+    return way.out_restriction.from == from && way.out_restriction.via == to;
+}
+
+std::size_t WayRestrictionMap::AsDuplicatedNodeID(const std::size_t restriction_id) const
+{
+    return std::distance(duplicated_node_groups.begin(),
+                         std::upper_bound(duplicated_node_groups.begin(),
```

Comment:
This should be split to two expression for readability, the `1` is almost sneaking away at the end.